client/packages/coldchain/src/Equipment/ImportAsset/UploadTab.tsx

```diff
@@ -0,0 +1,173 @@
+import React, { FC, useState } from 'react';
+import Papa from 'papaparse';
+import { ImportPanel } from './ImportPanel';
+import { useNotification } from '@common/hooks';
+import { InlineProgress, Typography, Upload } from '@common/components';
+import { useTranslation } from '@common/intl';
+import {
+  Grid,
+  Stack,
+  Paper,
+  Link,
+  FnUtils,
+  FileUtils,
+  InsertAssetInput,
+} from '@openmsupply-client/common';
+import * as EquipmentImportModal from './EquipmentImportModal';
+import { ImportRow, toInsertEquipmentInput } from './EquipmentImportModal';
+import { importEquipmentToCsv } from '../utils';
+import { AssetCatalogueItemFragment } from '@openmsupply-client/system';
+interface EquipmentUploadTabProps {
+  setEquipment: React.Dispatch<React.SetStateAction<ImportRow[]>>;
+  setErrorMessage: (value: React.SetStateAction<string>) => void;
+  onUploadComplete: () => void;
+  catalogueItemData?: AssetCatalogueItemFragment[];
+}
+
+export const EquipmentUploadTab: FC<ImportPanel & EquipmentUploadTabProps> = ({
+  tab,
+  setErrorMessage,
+  setEquipment,
+  onUploadComplete,
+  catalogueItemData,
+}) => {
+  const t = useTranslation('coldchain');
+  const { error } = useNotification();
+  const [isLoading, setIsLoading] = useState(false);
+  const EquipmentBuffer: EquipmentImportModal.ImportRow[] = [];
+
+  const csvExample = async () => {
+    const emptyRows: ImportRow[] = [];
+    const csv = importEquipmentToCsv(
+      emptyRows.map((row: ImportRow): InsertAssetInput => {
+        return toInsertEquipmentInput(row, catalogueItemData);
+      }),
+      t
+    );
+    FileUtils.exportCSV(csv, t('filename.cce'));
+  };
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const csvImport = (files: any) => {
+    setErrorMessage('');
+    EquipmentBuffer.length = 0; // Reset the import buffer
+    const csvFile = files[0];
+
+    if (!csvFile.name.endsWith('.csv')) {
+      setErrorMessage(t('messages.invalid-file'));
+      return;
+    }
+
+    if (csvFile) {
+      setIsLoading(true);
+      Papa.parse(csvFile, {
+        header: true,
+        worker: true,
+        skipEmptyLines: true,
+        chunkSize: 100 * 1024, // 100kb
+        chunk: processUploadedDataChunk,
+        complete: () => {
+          setEquipment(EquipmentBuffer);
+          setIsLoading(false);
+          onUploadComplete();
+        },
+      });
+    }
+    error(t('messages.error-no-file-selected'));
+  };
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  const processUploadedDataChunk = (data: any) => {
```

Comment:
```suggestion
  const processUploadedDataChunk = <T extends File>(data: ParseResult<T>) => {
```

just change the papaparse import:

```
import Papa, { ParseResult } from 'papaparse';
```

Reply:
I've gone down a bit of a chain of fixes to explicitly handle the dynamic keys of the parsed result.

It also exposes some weird double checking for undefined behaviour from our linter...

Maybe OK? Curious to hear your thoughts. I find these type definitions a bit arcane, so I'm unsure of the best approach here.

Fixes are:

```
// introduce new interface to accomomdate dynamic keys of parsed result
interface ParsedAsset {
  id: string;
  [key: string]: string | undefined;
}
```

and

```
const processUploadedDataChunk = (data: ParseResult<ParsedAsset>) => {
```

and here double checking for undefined (linter complains with just a single check?)

 ```
 if (
        row[t('label.asset-number')] &&
        row[t('label.asset-number')]?.trim() != ''
      ) {
        importRow.assetNumber = row[t('label.asset-number')] ?? '';
```

Reply:
The dynamic indexing of the row object is a bit problematic. it's a shame that the row props are the headings which are translatable. 

How about [this suggestion](https://github.com/msupply-foundation/open-msupply/compare/3139-import-asset-from-spreadsheet...3139-import-asset-from-spreadsheet(enum-version)) as a bit of a column number attempt?

I've used an enum to specify the column indices and extracted the row keys so that we can use those. An alternative version would be to specify the translations, like:

```
    const columns = {
      AssetNumber: t('label.asset-number'),
      CatalogueItemCode: t('label.catalogue-item-code'),
      Notes: t('label.asset-notes'),
    }

    const assetNumber = row(columns.AssetNumber);
```

which makes the code neater and doesn't rely on the ordering of the keys. The disadvantage is that you lose the ability for users to name columns as they want, as long as the columns are in the correct order they'll import ok..

Reply:
I imagine that having dynamic translations would be important given this is user facing. 

Is that a fair assumption?

I'll give your suggestion a go

Reply:
If it means compromising on code quality, I'd drop the requirement. It's good to have, but not essential